### PR TITLE
Compiler fix aggregations and functions expressions

### PR DESCRIPTION
**Pull request by bachdavi**
I left the `args` in the defrecord of the aggregation as we might need it in the future for multiple aggregations in the same :find clause.
I just realised that that could also be done via comparing the variables :Aggregation has been called with and the key symbols.

### CODE REVIEW

File: src/clj_3df/compiler.cljc [hunk 1]
```diff
@@ -550,7 +563,7 @@
         aggregation (->> (:find ir)
                          extract-aggregations
                          (reduce (fn [unified {:keys [aggregation-fn vars]}]
-                                   (->Aggregation aggregation-fn vars unified)) unified))
+                                   (->Aggregation aggregation-fn vars (extract-key-symbols (:find ir)) unified (extract-find-symbols (:find ir)))) unified))
```
Comment (comnik) [inline src/clj_3df/compiler.cljc]: Looks like `extract-key-symbols` could be done once, outside of the reduction?

File: src/clj_3df/compiler.cljc [hunk 2]
```diff
@@ -240,10 +240,10 @@
   (bound-symbols [this]
     (if (some? binding) (bound-symbols binding) args))
```
Comment (comnik) [inline src/clj_3df/compiler.cljc:241]: Unrelated to this PR, but this probably is not fully correct yet. At the very least, `result-sym` should be part of the bound symbols. But we also have to think about whether the ouput binding is supposed to replace the input binding, or whether they co-exist, etc...

File: src/clj_3df/compiler.cljc [hunk 3]
```diff
@@ -496,12 +505,16 @@
...
 ;; @TODO compose predicate functions
 (defmethod unify [Predicate Predicate] [p1 p2] (update p1 :binding unify p2))
+(defmethod unify [Predicate FnExpr] [pred f] (unify f pred))
 
 (defmethod unify [::binding FnExpr] [b1 b2] (unify b2 b1))
 (defmethod unify [FnExpr ::binding] [f binding] (update f :binding unify binding))
 (defmethod unify [FnExpr Projection] [b1 b2] (unify b2 b1))
 (defmethod unify [Projection FnExpr] [proj f] (update proj :binding unify f))
 (defmethod unify [FnExpr FnExpr] [p1 p2] (update p1 :binding unify p2))
+(defmethod unify [FnExpr Predicate] [f pred] (update f :binding unify pred))
```
Comment (comnik) [inline src/clj_3df/compiler.cljc:515]: I think we should throw here for now. Should never happen in practice, as long as output bindings are distinct from input bindings for function expressions.

Comment (bachdavi) [reply src/clj_3df/compiler.cljc:515]: What if the query includes predicate and function expressions?
Like so:
```
:find ?t ?t-2
:in $ ?cut-off
:where
[?e :time ?t]
[(interval ?t) ?d]
[(> ?t ?cut-off)]
```
The unify method will need to unify a function on the left and a predicate on the right.

File: src/clj_3df/compiler.cljc (hunk 1 above)
Comment (bachdavi) [reply src/clj_3df/compiler.cljc]: True.
Why not ditch the whole reduce of extract-aggregation and not allow multiple aggregations in the :find clause for now, until we have a better backend representation.
The way we have it implemented leads to nested `:Aggregation` calls which don't make sense in the backend when the dataflow is constructed.
We feed the output of one aggregation into another which leads in the best case to error and in the worst to panics.
Multiple aggregations would need  a duplication and subsequent joining of the underlying dataflows create by the bindings.
Or am I missing something?

File: src/clj_3df/compiler.cljc (hunk 3 above)
Comment (comnik) [reply src/clj_3df/compiler.cljc:515]: Fair!

File: src/clj_3df/compiler.cljc (hunk 2 above)
Comment (bachdavi) [reply src/clj_3df/compiler.cljc:241]: Yep good point. I'll have a look.

File: src/clj_3df/compiler.cljc (hunk 1 above)
Comment (comnik) [reply src/clj_3df/compiler.cljc]: Fine with me. We can look at something like `(->Aggregation [fn vars ...] [fn vars ...] ...+)`. Should also allow for the most efficient implementations (e.g. re-use the same grouping and compute multiple aggregates per group).

But removing the reduce for now and throwing on more than one aggregation makes sense for now.

File: src/clj_3df/compiler.cljc (hunk 2 above)
Comment (bachdavi) [reply src/clj_3df/compiler.cljc:241]: I'll create an issue.

**Pull request merged**